### PR TITLE
example: add GSI DynamoDB adapter

### PR DESCRIPTION
**Pull request by SachinShekhar**
DynamoDB Adapter to be used by the library.

### CODE REVIEW

File: example/adapters/dynamodb.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+      KeyConditionExpression: "grantId = :grantId",
+      ExpressionAttributeValues: {
+        ":grantId": grantId,
+      },
+      ProjectionExpression: "modelId",
+    };
+
+    const results = <{ modelId: string }[] | undefined>(
+      (await dynamoClient.query(params).promise()).Items
+    );
```
[the remark this answers is not in the document]
Comment (SachinShekhar) [reply example/adapters/dynamodb.ts:191]: There's no point of pagination on server. Multiple queries would increase latency. One query can return upto 1MB of data which is enough for practical purpose here.

[the remark this answers is not in the document]
Comment (panva) [reply example/adapters/dynamodb.ts:191]: So its unlikely LastEvaluatedKey will ever pop up and additional queries and delay would occur. Nevertheless having the code in place is the right thing here.

[the remark this answers is not in the document]
Comment (SachinShekhar) [reply example/adapters/dynamodb.ts]: Additional queries carry extra network overhead even after turning ON SDK's keep alive feature. But, I've updated the code anyway. Maybe, for most use cases that won't matter much.

**Comment by SachinShekhar** [comment]
> Generally I prefer to have LSI based dynamo implementations where each model is its own bucket

LSI can't replace GSI here. LSI can provide more choice of Sort Key, but the original Partition Key will still be required to access/ query data.

I can remove dependence on indexes fully by making keys complex & duplicating data (adjacency list pattern is one option), but that will be an overkill for such a simple use case which has no complex relations, not to mention it'll create complete mess in the db which will give no sense of models.

In the current form, with GSI, each model is indeed a virtual bucket. I've made sure to prefix all modelId with model name. So, if you want to see all items that belong to a model, you can query all items with PartitionKey starting with the model name. Everything is transparent, simple and scalable.

Another option is creation of different tables for different models, but that's an anti-pattern for DynamoDB unless a use case is justified like hotspot in db triggered by time series data.

**Comment by panva** [comment]
LSI based adapter has its benefits - cost and possible consistent reads, and yes - it can absolutely replace gsi. There are always tradeoffs. All i asked is that the filename in your PR is named dynamodb-gsi.ts ;) did not mean to start a discussion of lsi vs gsi.

**Comment by SachinShekhar** [comment]
> All i asked is that the filename in your PR is named dynamodb-gsi.ts ;)

LOL. Done!

**Comment by SachinShekhar** [comment]
BTW, a non-GSI data modeling doesn't need LSI. As I mentioned before, it can be done solely on its own.

Talking about cost, it's debatable. GSI's throughput price is high, but a non-GSI setup can still cost more because it'll need more operations to do same thing.

Talking about eventual vs strong consistency, do we even need strong consistency here in this case? What are the odds of multiple operations targeting same item here?

**Comment by panva** [comment]
Thank you for the contribution, i'm sure someone's going to find it useful.

> BTW, a non-GSI data modeling doesn't need LSI. As I mentioned before, it can be done solely on its own.
> 
> Talking about cost, it's debatable. GSI's throughput price is high, but a non-GSI setup can still cost more because it'll need more operations to do same thing.
> 
> Talking about eventual vs strong consistency, do we even need strong consistency here in this case? What are the odds of multiple operations targeting same item here?

It all depends on the application. In banking, and i do have some feedback from those applications, strong consistency is key. In such application you can live with the LSI 10GiB per shard limitation since tokens and sessions only live for very short time but you need a lot of throughput and strong consistency, especially when session bound tokens are in use. Scaling is also waaay easier, you just scale one table and it applies to your secondary indexes. And yet you have a clean DB "schema".

So again, application is everything.

**Pull request merged**